### PR TITLE
Impl fluentd file buffer.

### PR DESCRIPTION
**Pull request by jcantrill**
This PR removes memory buffer for fluentd and supercedes #525

### CODE REVIEW

**Comment by richm** [comment]
[test]

**Comment by jcantrill** [comment]
Ansible companion PR: https://github.com/openshift/openshift-ansible/pull/4851

File: fluentd/configs.d/openshift/es-copy-config.conf [hunk 1]
```diff
@@ -20,6 +20,8 @@
       flush_interval 5s
       max_retry_wait 300
       disable_retry_limit true
+      buffer_type file
+      buffer_path '/var/lib/fluentd/buffer'
```
Comment (richm) [inline fluentd/configs.d/openshift/es-copy-config.conf]: The error is
```
config error file="/etc/fluent/fluent.conf" error="Other 'elasticsearch_dynamic' plugin already use same buffer_path: type = elasticsearch_dynamic, buffer_path = /var/lib/fluentd/buffer"
```
I think this means that every output plugin must use a different path.  Perhaps use
```
buffer_path '/var/lib/fluentd/buffer-es-copy'
....
buffer_path '/var/lib/fluentd/buffer-es-ops-copy'
...
buffer_path '/var/lib/fluentd/buffer-es'
...
buffer_path '/var/lib/fluentd/buffer-es-ops'
etc.
```
and the secure_forward will also need a unique buffer_path

**Comment by jcantrill** [comment]
@richm rebased and pushed updates.  I'll work on a test run now

**Comment by jcantrill** [comment]
Locally this passes if tested with the changes in ansible.  I assume that may be what kills the tests on the CI machine.

File: fluentd/run.sh [hunk 2]
```diff
@@ -129,32 +125,68 @@ else
...
+    if [ "$ES_COPY" = "true" ]; then
+        DIV=`expr $DIV \* 2`
+    fi
 fi
 
-DIV=1
-if [ "$ES_HOST" != "$OPS_HOST" ] || [ "$ES_PORT" != "$OPS_PORT" ] ; then
-    # using ops cluster
-    DIV=`expr $DIV \* 2`
+FILE_BUFFER_PATH=/var/lib/fluentd/buffer
```
Comment (richm) [inline fluentd/run.sh]: This should be `FILE_BUFFER_PATH=/var/lib/fluentd/` ?

File: fluentd/run.sh [hunk 3]
```diff
@@ -129,32 +125,68 @@ else
...
+if [ "${USE_MUX_CLIENT:-}" = "true" ] ; then
+    # mux client has just one output.
+    DIV=1
 else
-    TOTAL_MEMORY_LIMIT=`echo $FLUENTD_MEMORY_LIMIT |  sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc`
-fi
-BUFFER_SIZE_LIMIT=`echo $BUFFER_SIZE_LIMIT |  sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc`
-if [ $BUFFER_SIZE_LIMIT -eq 0 ]; then
-    BUFFER_SIZE_LIMIT=1048576
+    # fluend usually has 2 putputs.
```
Comment (portante) [inline fluentd/run.sh]: putputs!  I love the double meaning.  =)

File: fluentd/run.sh [hunk 4]
```diff
@@ -129,32 +125,68 @@ else
 fi
 export K8S_FILTER_REMOVE_KEYS
 
-if [ "${USE_MUX:-}" = "true" ] ; then
-    TOTAL_MEMORY_LIMIT=`echo $MUX_MEMORY_LIMIT |  sed -e "s/[Kk]/*1024/g;s/[Mm]/*1024*1024/g;s/[Gg]/*1024*1024*1024/g;s/i//g" | bc`
+# How many outputs?
+if [ "${USE_MUX_CLIENT:-}" = "true" ] ; then
+    # mux client has just one output.
```
Comment (portante) [inline fluentd/run.sh]: Could we consider changing this to say, "A fluentd collector configured as mux client has just one output: sending to a mux".

File: fluentd/run.sh (hunk 2 above)
Comment (nhosoi) [inline fluentd/run.sh]: +1

File: fluentd/run.sh [hunk 5]
```diff
@@ -129,32 +125,68 @@ else
...
+    echo "ERROR: Invalid file buffer limit ($FILE_BUFFER_LIMIT) is given.  Failed to convert to bytes."
+    exit 1
+fi
+# Available disk space is less than the specified size.
+if [ $DF_LIMIT -lt $TOTAL_LIMIT ]; then
+    echo "WARNING: Available disk space ($DF_LIMIT bytes) is less than the user specified file buffer limit ($FILE_BUFFER_LIMIT)."
+    TOTAL_LIMIT=$DF_LIMIT
+fi
+
+if [ -z $TOTAL_LIMIT -o $TOTAL_LIMIT -eq 0 ]; then
```
Comment (portante) [inline fluentd/run.sh]: Why do we need this check?  Doesn't the check at line 160 ensure `$TOTAL_LIMIT` is strictly greater than 0?

File: fluentd/run.sh [hunk 6]
```diff
@@ -129,32 +125,68 @@ else
...
+    if [ "$USE_MUX" = "true" ]; then
+        # 75 % of 2G
+        TOTAL_LIMIT=1610612736
```
Comment (portante) [inline fluentd/run.sh]: Is this making the assumption that this default size is small enough that we don't care how much space is taken up on the volume?  How do we sure this value won't fill up the disk on which the buffer is located?

File: hack/testing/init-log-stack [hunk 7]
```diff
@@ -11,16 +11,38 @@ ES_HOSTNAME=${ES_HOST:+openshift_logging_es_hostname=$ES_HOST}
...
+kind: \"PersistentVolume\"
+metadata:
+  name: logging-muxpv-1
+spec:
+  capacity:
+    storage: \"6Gi\"
+  accessModes:
+    - \"ReadWriteOnce\"
+  hostPath:
+    path: ${FILE_BUFFER_PATH:-/var/log/mux/filebufferstorage}" > /tmp/host-pv.yaml
```
Comment (portante) [inline hack/testing/init-log-stack]: Does the path have to be /var/lib/fluentd?

File: fluentd/run.sh (hunk 5 above)
Comment (nhosoi) [inline fluentd/run.sh]: A good catch!  The check at line 160 was added based on a review.  This check at 170 should have been removed together.

File: fluentd/run.sh (hunk 6 above)
Comment (nhosoi) [inline fluentd/run.sh]: This part should be removed.

File: hack/testing/init-log-stack (hunk 7 above)
Comment (nhosoi) [inline hack/testing/init-log-stack]: +1

**Comment by nhosoi** [comment]
Note: with openshift-ansible pr/4851 and pr/4879, this pr 544 makes the pvc part of CI test passes ok.

**Comment by nhosoi** [comment]
Without the openshift-ansible PRs, I found the CI test failed unless the following  MUX_FILE_BUFFER_STORAGE_TYPE initialization.  @jcantrill, could you please add the line to test-mux.sh?
```
diff --git a/hack/testing/test-mux.sh b/hack/testing/test-mux.sh
index 26ce741..556e5b4 100755
--- a/hack/testing/test-mux.sh
+++ b/hack/testing/test-mux.sh
@@ -346,6 +346,7 @@ if [ -z "`get_running_pod es`" ] ; then
     exit 1
 fi
 
+MUX_FILE_BUFFER_STORAGE_TYPE=${MUX_FILE_BUFFER_STORAGE_TYPE:-emptydir}
 if [ "$MUX_FILE_BUFFER_STORAGE_TYPE" = "pvc" ]; then
     echo file_buffer_storage_type: pvc >> $MUXDEBUG
     oc get pv >> $MUXDEBUG
```

**Comment by portante** [comment]
@nhosoi, @jcantrill: can we merge this PR with #556?  If you need help with git operations to get these all back together, please let me know.  As it is, these requests do not appear to be ready to be merged.

**Comment by portante** [comment]
Okay, it looks like #556 is completely superseding this one.

**Pull request closed without merging**